server/datastore/mysql/apple_mdm.go

```diff
@@ -1265,79 +1284,92 @@ WHERE
 		AND ds.profile_id IS NULL AND ds.uuid IS NULL
 		-- except "remove" operations in any state
 		AND ( hmap.operation_type IS NULL OR hmap.operation_type != ? )
-		-- profiles that are being installed
-	`
-
-		inArgs := []any{uuids, uuids, fleet.MDMAppleOperationTypeRemove}
-		if len(installIdentifiers) > 0 {
-			profilesToRemoveStmt += `AND hmap.profile_identifier NOT IN (?)`
-			inArgs = append(inArgs, installIdentifiers)
-
-		}
+		`
 
-		stmt, args, err = sqlx.In(profilesToRemoveStmt, inArgs...)
+		stmt, args, err = sqlx.In(profilesToRemoveStmt, uuids, uuids, fleet.MDMAppleOperationTypeRemove)
 		if err != nil {
 			return ctxerr.Wrap(ctx, err, "building profiles to remove statement")
 		}
 		var profilesToRemove []*fleet.MDMAppleProfilePayload
 		err = sqlx.SelectContext(ctx, tx, &profilesToRemove, stmt, args...)
 		if err != nil {
-			return ctxerr.Wrap(ctx, err, "bulk set pending profile status execute")
+			return ctxerr.Wrap(ctx, err, "fetching profiles to remove")
 		}
 
 		if len(profilesToInstall) == 0 && len(profilesToRemove) == 0 {
 			return nil
 		}
 
-		// before doing the inserts, remove profiles with identifiers that will be re-sent
-		if len(profilesToInstall) > 0 {
-			var dargs []any
-			var dsb strings.Builder
-			for identifier, hostUUIDs := range identifierToHosts {
-				for _, hostUUID := range hostUUIDs {
-					dargs = append(dargs, hostUUID, identifier)
-					dsb.WriteString("(?,?),")
-				}
-			}
-			stmt = fmt.Sprintf(`DELETE FROM host_mdm_apple_profiles WHERE (host_uuid, profile_identifier) IN(%s)`, strings.TrimSuffix(dsb.String(), ","))
-			_, err = tx.ExecContext(ctx, stmt, dargs...)
-			if err != nil {
-				return ctxerr.Wrap(ctx, err, "bulk set pending profile status execute")
-			}
+		// delete all host profiles to start from a clean slate, new entries will be added next
+		// TODO(roberto): is this really necessary? this was pre-existing
+		// behavior but I think it can be refactored. For now leaving it as-is.
+		if err := bulkDeleteMDMAppleHostsConfigProfilesDB(ctx, tx, profilesToInstall); err != nil {
+			return err
 		}
 
+		// keptProfiles tracks profilesToAdd ∩ profilesToRemove, this is used to avoid:
+		//
+		// - Sending a RemoveProfile followed by an InstallProfile for a
+		// profile with an identifier that's already installed, which can cause
+		// racy behaviors.
+		// - Sending a InstallProfile command for a profile that's exactly the
+		// same as the one installed. Customers have reported that sending the
+		// command causes unwanted behavior.
+		keptProfiles := apple_mdm.NewProfileBimap()
```

Comment:
Is there a reason to be using bimap here at the datastore level and also in `ReconcileProfiles` at the service level?

Reply:
the code in these two functions is not very DRY, they do almost the same thing but get called from different places. We should refactor that at some point